File: packages/typescript-sdk/src/authoring/test-client.ts
import type { RunContext } from "./job";
import { createRunContext, type RunContextClient } from "./run-context";

export type TestRunRecord = {
  readonly checkpoints: Record<string, unknown>[];
  readonly logs: {
    level: string;
    message: string;
    data?: Record<string, unknown>;
  }[];
  readonly usageReports: {
    provider: string;
    model: string;
    promptTokens?: number;
    completionTokens?: number;
    totalTokens?: number;
    costMicrousd?: number;
  }[];
  readonly toolCalls: {
    toolName: string;
    input?: Record<string, unknown>;
    output?: Record<string, unknown>;
    durationMs?: number;
    status?: string;
  }[];
  readonly outputs: { key: string; value: Record<string, unknown> }[];
  readonly progressUpdates: { percent: number; message?: string }[];
  readonly stateStore: Map<string, unknown>;
  readonly streamChunks: {
    chunk: string;
    streamId?: string;
    done?: boolean;
  }[];
  heartbeats: number;
  readonly spawns: {
    jobSlug: string;
    projectId: string;
    payload?: Record<string, unknown>;
  }[];
  readonly events: {
    eventKey: string;
    timeoutSecs?: number;
    notifyUrl?: string;
  }[];
  readonly annotations: Record<string, string>[];
  readonly continuations: { payload?: Record<string, unknown> }[];
  completed: boolean;
  failed: boolean;
  failError?: string;
  result?: Record<string, unknown>;
};

export const createTestContext = (
  runId = "test-run",
  options?: { attempt?: number; signal?: AbortSignal }
): { ctx: RunContext; record: TestRunRecord } => {
  const record: TestRunRecord = {
    checkpoints: [],
    logs: [],
    usageReports: [],
    toolCalls: [],
    outputs: [],
    progressUpdates: [],
    stateStore: new Map(),
    streamChunks: [],
    heartbeats: 0,
    spawns: [],
    events: [],
    annotations: [],
    continuations: [],
    completed: false,
    failed: false,
  };

  const mockClient: RunContextClient = {
    checkpointRun: ({ body }) => {
      record.checkpoints.push(body.state);
      return Promise.resolve();
    },
    heartbeatRun: () => {
      record.heartbeats++;
      return Promise.resolve();
    },
    progressRun: ({ body }) => {
      record.progressUpdates.push({
        percent: body.percent,
        message: body.message,
      });
      return Promise.resolve();
    },
    logRun: ({ body }) => {
      record.logs.push({
        level: body.level,
        message: body.message,
        data: body.data,
      });
      return Promise.resolve();
    },
    usageRun: ({ body }) => {
      record.usageReports.push({
        provider: body.provider,
        model: body.model,
        promptTokens: body.prompt_tokens,
        completionTokens: body.completion_tokens,
        totalTokens: body.total_tokens,
        costMicrousd: body.cost_microusd,
      });
      return Promise.resolve();
    },
    toolCallRun: ({ body }) => {
      record.toolCalls.push({
        toolName: body.tool_name,
        input: body.input,
        output: body.output,
        durationMs: body.duration_ms,
        status: body.status,
      });
      return Promise.resolve();
    },
    outputRun: ({ body }) => {
      record.outputs.push({ key: body.key, value: body.value });
      return Promise.resolve();
    },
    waitForEventRun: ({ body }) => {
      record.events.push({
        eventKey: body.event_key,
        timeoutSecs: body.timeout_secs,
        notifyUrl: body.notify_url,
      });
      return Promise.resolve({
        status: "waiting",
        eventKey: body.event_key,
        triggerId: "trigger_test",
        expiresAt: new Date(
          Date.now() + (body.timeout_secs ?? 300) * 1000
        ).toISOString(),
      });
    },
    spawnRun: ({ body }) => {
      record.spawns.push({
        jobSlug: body.job_slug,
        projectId: body.project_id,
        payload: body.payload,
      });
      return Promise.resolve({ id: `spawn_${record.spawns.length}` });
    },
    continueRun: ({ body }) => {
      record.continuations.push({
        payload: body?.payload as Record<string, unknown> | undefined,
      });
      return Promise.resolve({
        id: `continue_${record.continuations.length}`,
      });
    },
    annotateRun: ({ body }) => {
      record.annotations.push(body.annotations);
      return Promise.resolve();
    },
    completeRun: ({ body }) => {
      record.completed = true;
      record.result = body?.result as Record<string, unknown> | undefined;
      return Promise.resolve();
    },
    failRun: ({ body }) => {
      record.failed = true;
      record.failError = body.error;
      return Promise.resolve();
    },
    stateRun: ({ body }) => {
      record.stateStore.set(body.key, body.value);
      return Promise.resolve();
    },
    getStateByRunId: () =>
      Promise.resolve(
        Array.from(record.stateStore.entries()).map(([key, value]) => ({
          key,
          value,
          updatedAt: new Date().toISOString(),
        }))
      ),
    getStateByRunIdAndKey: ({ pathParams }) =>
      Promise.resolve(record.stateStore.get(pathParams.key)),
    deleteState: ({ pathParams }) => {
      record.stateStore.delete(pathParams.key);
      return Promise.resolve();
    },
    streamRun: ({ body }) => {
      record.streamChunks.push({
        chunk: body.chunk,
        streamId: body.stream_id,
        done: body.done,
      });
      return Promise.resolve();
    },
  };

  return { ctx: createRunContext(mockClient, runId, { attempt: options?.attempt, signal: options?.signal }), record };
};

File: packages/python-sdk/strait/authoring/_test_client.py
"""In-memory test context for unit-testing jobs and agents without HTTP."""

from __future__ import annotations

from dataclasses import dataclass, field
from datetime import datetime, timezone
from typing import Any

from strait.authoring._run_context import RunContext
from strait.authoring._run_context_client import create_run_context


@dataclass
class TestRunRecord:
    checkpoints: list[dict[str, Any]] = field(default_factory=list)
    logs: list[dict[str, Any]] = field(default_factory=list)
    usage_reports: list[dict[str, Any]] = field(default_factory=list)
    tool_calls: list[dict[str, Any]] = field(default_factory=list)
    outputs: list[dict[str, Any]] = field(default_factory=list)
    progress_updates: list[dict[str, Any]] = field(default_factory=list)
    state_store: dict[str, Any] = field(default_factory=dict)
    stream_chunks: list[dict[str, Any]] = field(default_factory=list)
    heartbeats: int = 0
    spawns: list[dict[str, Any]] = field(default_factory=list)
    events: list[dict[str, Any]] = field(default_factory=list)
    annotations: list[dict[str, str]] = field(default_factory=list)
    continuations: list[dict[str, Any]] = field(default_factory=list)
    completed: bool = False
    failed: bool = False
    fail_error: str | None = None
    result: dict[str, Any] | None = None


def create_test_context(
    run_id: str = "test-run",
    attempt: int = 1,
) -> tuple[RunContext, TestRunRecord]:
    record = TestRunRecord()

    class MockClient:
        async def checkpoint_run(self, rid: str, body: Any) -> None:
            record.checkpoints.append(body.get("state", body))

        async def heartbeat_run(self, rid: str) -> None:
            record.heartbeats += 1

        async def progress_run(self, rid: str, body: Any) -> None:
            record.progress_updates.append(body)

        async def log_run(self, rid: str, body: Any) -> None:
            record.logs.append(body)

        async def usage_run(self, rid: str, body: Any) -> None:
            record.usage_reports.append(body)

        async def tool_call_run(self, rid: str, body: Any) -> None:
            record.tool_calls.append(body)

        async def output_run(self, rid: str, body: Any) -> None:
            record.outputs.append(body)

        async def wait_for_event_run(self, rid: str, body: Any) -> dict[str, Any]:
            record.events.append(body)
            return {
                "status": "waiting",
                "event_key": body["event_key"],
                "trigger_id": "trigger_test",
                "expires_at": datetime.now(timezone.utc).isoformat(),
            }

        async def spawn_run(self, rid: str, body: Any) -> dict[str, Any]:
            record.spawns.append(body)
            return {"id": f"spawn_{len(record.spawns)}"}

        async def continue_run(self, rid: str, body: Any) -> dict[str, Any]:
            payload = body.get("payload") if body else None
            record.continuations.append({"payload": payload})
            return {"id": f"continue_{len(record.continuations)}"}

        async def annotate_run(self, rid: str, body: Any) -> None:
            record.annotations.append(body.get("annotations", body))

        async def complete_run(self, rid: str, body: Any) -> None:
            record.completed = True
            if body:
                record.result = body.get("result")

        async def fail_run(self, rid: str, body: Any) -> None:
            record.failed = True
            record.fail_error = body.get("error")

        async def set_state(self, rid: str, body: Any) -> None:
            record.state_store[body["key"]] = body["value"]

        async def list_state(self, rid: str) -> list[dict[str, Any]]:
            return [
                {"key": k, "value": v, "updated_at": datetime.now(timezone.utc).isoformat()}
                for k, v in record.state_store.items()
            ]

        async def get_state(self, rid: str, key: str) -> Any:
            return record.state_store.get(key)

        async def delete_state(self, rid: str, key: str) -> None:
            record.state_store.pop(key, None)

        async def stream_run(self, rid: str, body: Any) -> None:
            record.stream_chunks.append(body)

    ctx = create_run_context(MockClient(), run_id, attempt)
    return ctx, record
